shell/platform/windows/text_input_manager.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/text_input_manager.h"
+
+#include <imm.h>
+#include <memory>
+
+namespace flutter {
+
+void TextInputManager::CreateImeWindow(HWND hwnd) {
+  // Some IMEs ignore calls to ::ImmSetCandidateWindow() and use the position of
+  // the current system caret instead via ::GetCaretPos(). In order to behave
+  // as expected with these IMEs, we create a temporary system caret.
+  if (!ime_active_) {
+    ::CreateCaret(hwnd, nullptr, 1, 1);
+  }
+  ime_active_ = true;
+
+  // Set the position of the IME windows.
+  UpdateImeWindow(hwnd);
+}
+
+void TextInputManager::DestroyImeWindow(HWND hwnd) {
+  // Destroy the system caret created in CreateImeWindow().
+  if (ime_active_) {
+    ::DestroyCaret();
+  }
+  ime_active_ = false;
+}
+
+void TextInputManager::UpdateImeWindow(HWND hwnd) {
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    MoveImeWindow(hwnd, imm_context);
+    ::ImmReleaseContext(hwnd, imm_context);
+  }
+}
+
+void TextInputManager::UpdateCaretRect(HWND hwnd, const Rect& rect) {
+  caret_rect_ = rect;
+
+  // TODO(cbracken): wrap these in an RAII container.
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    MoveImeWindow(hwnd, imm_context);
+    ::ImmReleaseContext(hwnd, imm_context);
+  }
+}
+
+long TextInputManager::GetComposingCursorPosition(HWND hwnd) const {
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    // Read the cursor position within the composing string.
+    const int pos = ImmGetCompositionStringW(imm_context, GCS_CURSORPOS, nullptr, 0);
+    ::ImmReleaseContext(hwnd, imm_context);
+    return pos;
+  }
+  return -1;
+}
+
+bool TextInputManager::GetComposingString(HWND hwnd, std::u16string* result) const {
```

Comment:
For these three methods, prefer `std::optional` return to bool+outparam.

Reply:
Done.